packages/server/src/websocket/websocket.service.ts

```diff
@@ -698,4 +653,151 @@ export class WebsocketService {
+   *
+   * @param {TakeFromQueueDto} dto - Data transfer object containing query parameters.
+   * @returns {Promise<QueuedMessage[]>} - A promise that resolves to an array of queued messages.
+   */
+  private async takeMessagesWithLimit(dto: TakeFromQueueDto): Promise<QueuedMessage[]> {
+    const { connectionId, limit, recipientDid } = dto
+
+    this.logger.debug('[takeMessagesWithLimit] Method called with DTO:', dto)
+
+    try {
+      // Query MongoDB with the provided connectionId or recipientDid, and state 'pending'
+      const mongoMessages = await this.queuedMessage
+        .find({
+          $or: [{ connectionId }, { recipientKeys: recipientDid }],
+          state: 'pending',
+        })
+        .sort({ createdAt: 1 })
+        .limit(limit)
+        .select({ messageId: 1, encryptedMessage: 1, createdAt: 1 })
+        .lean()
+        .exec()
+
+      const mongoMappedMessages: QueuedMessage[] = mongoMessages.map((msg) => ({
+        id: msg.messageId,
+        receivedAt: msg.createdAt,
+        encryptedMessage: msg.encryptedMessage,
+      }))
+
+      this.logger.debug(
+        `[takeMessagesWithLimit] Fetched ${mongoMappedMessages.length} messages from MongoDB for connectionId ${connectionId}`,
+      )
+
+      // Retrieve messages from Redis
+      const redisMessagesRaw = await this.redis.lrange(`connectionId:${connectionId}:queuemessages`, 0, limit - 1)
+      const redisMessages: QueuedMessage[] = redisMessagesRaw.map((message) => {
+        const parsedMessage = JSON.parse(message)
+
+        // Map Redis data to QueuedMessage type
+        return {
+          id: parsedMessage.messageId,
+          receivedAt: new Date(parsedMessage.receivedAt),
+          encryptedMessage: parsedMessage.encryptedMessage,
+        }
+      })
+
+      this.logger.debug(
+        `[takeMessagesWithLimit] Fetched ${redisMessages.length} messages from Redis for connectionId ${connectionId}`,
+      )
+      // Combine messages from Redis and MongoDB
+      const combinedMessages: QueuedMessage[] = [...mongoMappedMessages, ...redisMessages]
+
+      this.logger.debug(
+        `[takeMessagesWithLimit] combinedMessages for connectionId ${connectionId}: ${combinedMessages}`,
+      )
+
+      return combinedMessages
+    } catch (error) {
+      this.logger.error('[takeMessagesWithLimit] Error retrieving messages from Redis and MongoDB:', {
+        connectionId,
+        error: error.message,
+      })
+      return []
+    }
+  }
+
+  /**
+   * Retrieves messages from both Redis and MongoDB up to a specified total size in bytes.
+   *
+   * @param {TakeFromQueueDto} dto - Data transfer object containing query parameters.
+   * @returns {Promise<QueuedMessage[]>} - A promise that resolves to an array of queued messages.
+   */
+  private async takeMessagesWithSize(dto: TakeFromQueueDto): Promise<QueuedMessage[]> {
+    const { connectionId, recipientDid, limitSize } = dto
+    const maxMessageSizeBytes = limitSize * 1024 * 1024
```

Comment:
Limit is expressed in bytes, since this is how it is specified in DIDComm constraints and other related specs (such as Pickup V2)

Reply:
Resolved!